addOns/frontendscanner/src/test/java/org/zaproxy/zap/extension/frontendscanner/FrontEndScannerProxyListenerUnitTest.java

```diff
@@ -0,0 +1,162 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpResponseHeader;
+import org.zaproxy.zap.testutils.TestUtils;
+
+/** Unit test for {@link FrontEndScannerProxyListener}. */
+public class FrontEndScannerProxyListenerUnitTest extends TestUtils {
+
+    private static final String HOSTNAME = "example.com";
+
+    FrontEndScannerProxyListener frontEndScannerProxyListener;
+    HttpMessage msg;
+
+    @BeforeEach
+    public void setUp() throws URIException, HttpMalformedHeaderException {
+        Logger rootLogger = Logger.getRootLogger();
+        rootLogger.setLevel(Level.OFF);
+
+        FrontEndScannerAPI api = mock(FrontEndScannerAPI.class);
+        FrontEndScannerOptions options = mock(FrontEndScannerOptions.class);
+        when(options.isEnabled()).thenReturn(true);
+
+        HistoryReference ref = mock(HistoryReference.class);
+        when(ref.getHistoryId()).thenReturn(42);
+
+        frontEndScannerProxyListener = new FrontEndScannerProxyListener(api, options);
+
+        msg = new HttpMessage(new URI("https", HOSTNAME, "/", ""));
+        msg.getResponseHeader().setHeader(HttpResponseHeader.CONTENT_TYPE, "text/html");
+        msg.setHistoryRef(ref);
+    }
+
+    @Test
+    public void testInjectTheFrontEndTrackerBeforeOtherScriptsInHeadTag() {
+        // Given
+        String htmlBody = "<head><script></script></head><body></body>";
+        msg.setResponseBody(htmlBody);
+
+        // When
+        frontEndScannerProxyListener.onHttpResponseReceive(msg);
+
+        // Then
+        String expectedHtmlFormat =
+                "<head><script src='https:\\/\\/"
+                        + HOSTNAME
+                        + "\\/zapCallBackUrl\\/-?[0-9]+\\?action=getFile&filename=front-end-scanner.js&historyReferenceId=42'><\\/script><script><\\/script><\\/head><body><\\/body>";
+        String result = msg.getResponseBody().toString();
+
+        assertTrue(result.matches(expectedHtmlFormat));
+    }
+
+    @Test
+    public void testInjectAfterMetaTagInHeadTag() {
+        // Given
+        String htmlBody = "<head><meta></head><body></body>";
+        msg.setResponseBody(htmlBody);
+
+        // When
+        frontEndScannerProxyListener.onHttpResponseReceive(msg);
+
+        // Then
+        String expectedHtmlFormat =
+                "<head><meta><script src='https:\\/\\/"
+                        + HOSTNAME
+                        + "\\/zapCallBackUrl\\/-?[0-9]+\\?action=getFile&filename=front-end-scanner.js&historyReferenceId=42'><\\/script><\\/head><body><\\/body>";
+        String result = msg.getResponseBody().toString();
+
+        assertTrue(result.matches(expectedHtmlFormat));
+    }
+
+    @Test
+    public void testRemovesCSPFromHttpResponsesIfInjecting() {
```

Comment:
- It would be more clear with a parameterized test, all headers would be tested regardless of one failing and the one failing would be more obviou.
 - Could you move to before (or after) the tests that are checking the injections?